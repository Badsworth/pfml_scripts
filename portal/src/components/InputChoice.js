import PropTypes from "prop-types";
import React from "react";
import classnames from "classnames";
import useUniqueId from "../hooks/useUniqueId";

/**
 * A checkbox or radio field. In most cases, you shouldn't use this
 * component directly. Instead, use `InputChoiceGroup` which renders
 * markup for the entire fieldset.
 *
 * [USWDS Reference â†—](https://designsystem.digital.gov/components/form-controls)
 */
function InputChoice({ type = "checkbox", ...props }) {
  const inputId = useUniqueId("InputChoice");
  const id = props.id || inputId;

  const fieldClassName = classnames(
    `usa-${type} measure-5 bg-transparent`,
    props.className
  );
  return (
    <div className={fieldClassName}>
      <input
        checked={props.checked}
        className={`usa-${type}__input`}
        disabled={props.disabled}
        id={id}
        name={props.name}
        onChange={props.onChange}
        type={type}
        value={props.value}
        aria-controls={props["aria-controls"]}
      />
      <label className={`usa-${type}__label`} htmlFor={id}>
        {props.label}

        {props.hint && (
          <span className="usa-hint text-base-dark">
            <br />
            {props.hint}
          </span>
        )}
      </label>
    </div>
  );
}

InputChoice.propTypes = {
  /**
   * HTML `aria-controls` attribute. Used to indicate that the input affects
   * another element.
   */
  "aria-controls": PropTypes.string,
  /**
   * Sets the input's `checked` state. Use this in combination with `onChange`
   * for a controlled component.
   */
  checked: PropTypes.bool,
  /**
   * Additional classes to include on the field
   */
  className: PropTypes.string,
  /**
   * HTML input `disabled` attribute
   */
  disabled: PropTypes.bool,
  /**
   * Unique identifier for input
   */
  id: PropTypes.string,
  /**
   * Localized hint text
   */
  hint: PropTypes.node,
  /**
   * Localized field label
   */
  label: PropTypes.node.isRequired,
  /**
   * HTML input `name` attribute
   */
  name: PropTypes.string.isRequired,
  /**
   * HTML input `onChange` attribute
   */
  onChange: PropTypes.func,
  /**
   * HTML input `type` attribute
   */
  type: PropTypes.oneOf(["checkbox", "radio"]),
  /**
   * Sets the input's `value`
   */
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
};

export default InputChoice;
