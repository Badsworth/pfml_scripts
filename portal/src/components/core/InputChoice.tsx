import React from "react";
import classnames from "classnames";
import isBlank from "../../utils/isBlank";
import useUniqueId from "../../hooks/useUniqueId";

interface Props {
  /**
   * HTML `aria-controls` attribute. Used to indicate that the input affects
   * another element.
   */
  "aria-controls"?: string;
  /**
   * Sets the input's `checked` state. Use this in combination with `onChange`
   * for a controlled component.
   */
  checked?: boolean;
  /**
   * Additional classes to include on the field
   */
  className?: string;
  /**
   * HTML input `disabled` attribute
   */
  disabled?: boolean;
  /**
   * Unique identifier for input
   */
  id?: string;
  /**
   * Localized hint text
   */
  hint?: React.ReactNode;
  /**
   * Localized field label
   */
  label: React.ReactNode;
  /**
   * HTML input `name` attribute
   */
  name: string;
  /**
   * HTML input `onChange` attribute
   */
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  /**
   * HTML input `type` attribute
   */
  type?: "checkbox" | "radio";
  /**
   * Sets the input's `value`
   */
  value: number | string;
}

/**
 * A checkbox or radio field. In most cases, you shouldn't use this
 * component directly. Instead, use `InputChoiceGroup` which renders
 * markup for the entire fieldset.
 *
 * [USWDS Reference â†—](https://designsystem.digital.gov/components/form-controls)
 */
function InputChoice({ type = "checkbox", ...props }: Props) {
  const inputId = useUniqueId("InputChoice");
  const id = props.id || inputId;

  const fieldClassName = classnames(
    `usa-${type} measure-5 bg-transparent`,
    props.className
  );
  return (
    <div className={fieldClassName}>
      <input
        checked={props.checked}
        className={`usa-${type}__input`}
        disabled={props.disabled}
        id={id}
        name={props.name}
        onChange={props.onChange}
        type={type}
        value={props.value}
        aria-controls={props["aria-controls"]}
      />
      <label className={`usa-${type}__label`} htmlFor={id}>
        {props.label}

        {!isBlank(props.hint) && (
          <span className="usa-hint text-base-dark">
            <br />
            {props.hint}
          </span>
        )}
      </label>
    </div>
  );
}

export default InputChoice;
