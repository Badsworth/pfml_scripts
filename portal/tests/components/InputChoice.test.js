import { render, screen } from "@testing-library/react";

import InputChoice from "../../src/components/InputChoice";
import React from "react";
import userEvent from "@testing-library/user-event";

describe("InputChoice", () => {
  const defaultProps = {
    label: "foo",
    name: "bar",
    value: "foobar",
  };

  describe("InputChoice attributes", () => {
    it("creates `id` attribute if `id` is not provided", () => {
      render(<InputChoice {...defaultProps} />);
      const input = screen.getByRole("checkbox");

      expect(input).toHaveAttribute("id");
      expect(input).toMatchSnapshot();
    });

    it("sets `aria-controls` attribute when provided", () => {
      render(<InputChoice {...defaultProps} ariaControls="foobar-aria" />);
      const input = screen.getByRole("checkbox");

      expect(input).toHaveAttribute("aria-controls");
      expect(input).toMatchSnapshot();
    });

    it("sets the `name` attribute when provided", () => {
      render(<InputChoice {...defaultProps} />);
      const input = screen.getByRole("checkbox");

      expect(input).toHaveAttribute("name", defaultProps.name);
      expect(input).toMatchSnapshot();
    });

    it("sets the `value` attribute when provided", () => {
      render(<InputChoice {...defaultProps} />);
      const input = screen.getByRole("checkbox");

      expect(input).toMatchInlineSnapshot(`
        <input
          class="usa-checkbox__input"
          id="InputChoice4"
          name="bar"
          type="checkbox"
          value="foobar"
        />
      `);
    });

    it("sets the `disabled` attribute when provided", () => {
      render(<InputChoice {...defaultProps} disabled />);
      const input = screen.getByRole("checkbox");

      expect(input).toBeDisabled();
      expect(input).toMatchSnapshot();
    });
  });

  describe("InputChoice class names", () => {
    it("sets the parent container's className when provided", () => {
      render(<InputChoice {...defaultProps} className="foo" />);
      const inputWrapper = screen.getByRole("checkbox").closest("div");

      expect(inputWrapper).toBeInTheDocument();
      expect(inputWrapper).toHaveClass("foo");
      expect(inputWrapper).toMatchSnapshot();
    });

    it("sets the correct default input class name", () => {
      render(<InputChoice {...defaultProps} />);
      const input = screen.getByRole("checkbox");

      expect(input).toHaveClass("usa-checkbox__input");
      expect(input).toMatchSnapshot();
    });
  });

  describe("InputChoice children", () => {
    it("renders the label element", () => {
      render(<InputChoice {...defaultProps} />);
      const label = screen.getByText(/foo/);

      expect(label).toHaveClass("usa-checkbox__label");
      expect(label).toBeInTheDocument();
      expect(label).toMatchSnapshot();
    });

    it("renders the hint when provided", () => {
      render(<InputChoice {...defaultProps} hint="Foobar Hint" />);
      const hint = screen.getByText(/foobar hint/i);

      expect(hint).toBeInTheDocument();
      expect(hint).toMatchSnapshot();
    });
  });

  describe("InputChoice `types`", () => {
    it("sets the correct default input type", () => {
      render(<InputChoice {...defaultProps} />);
      const input = screen.getByRole("checkbox");

      expect(input).toBeInTheDocument();
      expect(input).toMatchSnapshot();
    });

    it("sets input type to radio when `type` is `radio`", () => {
      render(<InputChoice {...defaultProps} type="radio" />);
      const input = screen.getByRole("radio");

      expect(input).toHaveClass("usa-radio__input");
      expect(input).toBeInTheDocument();
      expect(input).toMatchSnapshot();
    });
  });

  describe("InputChoice onChange", () => {
    it("calls `onChange` on click, when `onChange` is provided", () => {
      const handleChange = jest.fn();
      render(<InputChoice {...defaultProps} onChange={handleChange} />);
      const input = screen.getByRole("checkbox");
      userEvent.click(input);

      expect(handleChange).toHaveBeenCalled();
      expect(input).toMatchSnapshot();
    });
  });
});
