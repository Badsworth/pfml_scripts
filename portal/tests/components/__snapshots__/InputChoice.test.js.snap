// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputChoice InputChoice \`types\` sets input type to radio when \`type\` is \`radio\` 1`] = `
<input
  class="usa-radio__input"
  id="InputChoice11"
  name="bar"
  type="radio"
  value="foobar"
/>
`;

exports[`InputChoice InputChoice \`types\` sets the correct default input type 1`] = `
<input
  class="usa-checkbox__input"
  id="InputChoice10"
  name="bar"
  type="checkbox"
  value="foobar"
/>
`;

exports[`InputChoice InputChoice attributes creates \`id\` attribute if \`id\` is not provided 1`] = `
<input
  class="usa-checkbox__input"
  id="InputChoice1"
  name="bar"
  type="checkbox"
  value="foobar"
/>
`;

exports[`InputChoice InputChoice attributes sets \`aria-controls\` attribute when provided 1`] = `
<input
  aria-controls="foobar-aria"
  class="usa-checkbox__input"
  id="InputChoice2"
  name="bar"
  type="checkbox"
  value="foobar"
/>
`;

exports[`InputChoice InputChoice attributes sets the \`disabled\` attribute when provided 1`] = `
<input
  class="usa-checkbox__input"
  disabled=""
  id="InputChoice5"
  name="bar"
  type="checkbox"
  value="foobar"
/>
`;

exports[`InputChoice InputChoice attributes sets the \`name\` attribute when provided 1`] = `
<input
  class="usa-checkbox__input"
  id="InputChoice3"
  name="bar"
  type="checkbox"
  value="foobar"
/>
`;

exports[`InputChoice InputChoice children renders the hint when provided 1`] = `
<span
  class="usa-hint text-base-dark"
>
  <br />
  Foobar Hint
</span>
`;

exports[`InputChoice InputChoice children renders the label element 1`] = `
<label
  class="usa-checkbox__label"
  for="InputChoice8"
>
  foo
</label>
`;

exports[`InputChoice InputChoice class names sets the correct default input class name 1`] = `
<input
  class="usa-checkbox__input"
  id="InputChoice7"
  name="bar"
  type="checkbox"
  value="foobar"
/>
`;

exports[`InputChoice InputChoice class names sets the parent container's className when provided 1`] = `
<div
  class="usa-checkbox measure-5 bg-transparent foo"
>
  <input
    class="usa-checkbox__input"
    id="InputChoice6"
    name="bar"
    type="checkbox"
    value="foobar"
  />
  <label
    class="usa-checkbox__label"
    for="InputChoice6"
  >
    foo
  </label>
</div>
`;

exports[`InputChoice InputChoice onChange calls \`onChange\` on click, when \`onChange\` is provided 1`] = `
<input
  class="usa-checkbox__input"
  id="InputChoice12"
  name="bar"
  type="checkbox"
  value="foobar"
/>
`;
